File: src/traffic_light_classifier/cpp_version/src/cam_node.cpp
#include "traffic_light_classifier/cam_node.hpp"

#include <filesystem>
#include <fstream>
#include <sstream>

namespace traffic_light_classifier
{

CamNode::CamNode(const rclcpp::NodeOptions & options)
: Node("cam_node", options), gen_(rd_())
{
  publisher_ = this->create_publisher<sensor_msgs::msg::Image>(
    "/perception/traffic_light_recognition/traffic_light_image", 10);
  debug_publisher_ = this->create_publisher<traffic_light_msg::msg::TrafficLightMsg>(
    "/perception/traffic_light_recognition/traffic_light_data", 10);
  timer_ = this->create_wall_timer(
    std::chrono::seconds(1), std::bind(&CamNode::timerCallback, this));

  this->declare_parameter("annotations_dir", "/home/casia/jupyter_notebooks/Annotations/Annotations/dayTrain");
  this->declare_parameter("data_dir", "/home/casia/jupyter_notebooks/dayTrain/dayTrain");
  this->declare_parameter("debug_mode", false);

  // 新增参数
  this->declare_parameter("publish_cropped_images", false);

  annotations_dir_ = this->get_parameter("annotations_dir").as_string();
  data_dir_ = this->get_parameter("data_dir").as_string();
  
  // 获取参数值
  publish_cropped_images_ = this->get_parameter("publish_cropped_images").as_bool();

  RCLCPP_INFO(this->get_logger(), "Annotations directory: %s", annotations_dir_.c_str());
  RCLCPP_INFO(this->get_logger(), "Data directory: %s", data_dir_.c_str());
  RCLCPP_INFO(this->get_logger(), "Publish cropped images: %s", publish_cropped_images_ ? "true" : "false");

  loadDataset();

  RCLCPP_INFO(this->get_logger(), "Dataset size: %zu", annotations_.size());
  RCLCPP_INFO(this->get_logger(), "Classes: %s", getClassesString().c_str());
}

void CamNode::timerCallback()
{
  if (annotations_.empty()) {
    RCLCPP_ERROR(this->get_logger(), "No images loaded. Cannot publish.");
    return;
  }

  int idx = dis_(gen_);
  const auto& ann = annotations_[idx];

  cv::Mat image = cv::imread(ann.img_path);
  if (image.empty()) {
    RCLCPP_ERROR(this->get_logger(), "Failed to read image: %s", ann.img_path.c_str());
    return;
  }

  cv::Mat image_to_publish;

  if (publish_cropped_images_) {
    cv::Mat cropped_img = cropImage(image, ann.bbox);
    image_to_publish = cropped_img;
  } else {
    image_to_publish = image;
  }

  sensor_msgs::msg::Image::SharedPtr img_msg = 
    cv_bridge::CvImage(std_msgs::msg::Header(), "bgr8", image_to_publish).toImageMsg();
  
  publisher_->publish(*img_msg);
  
  if (this->get_parameter("debug_mode").as_bool()) {
    auto debug_msg = std::make_unique<traffic_light_msg::msg::TrafficLightMsg>();
    debug_msg->image = *img_msg;
    debug_msg->label = ann.label;
    debug_publisher_->publish(std::move(debug_msg));
    RCLCPP_INFO(this->get_logger(), "Published debug image with label: %s", ann.label.c_str());
  } else {
    RCLCPP_INFO(this->get_logger(), "Published image");
  }
}

// 其余代码保持不变

}  // namespace traffic_light_classifier

#include "rclcpp_components/register_node_macro.hpp"
RCLCPP_COMPONENTS_REGISTER_NODE(traffic_light_classifier::CamNode)

File: src/traffic_light_classifier/traffic_light_classifier/cam_node.py
import rclpy
from rclpy.node import Node
from sensor_msgs.msg import Image
from cv_bridge import CvBridge
import random
import os
import pandas as pd
import torch
from torch.utils.data import Dataset
from torchvision import transforms
from PIL import Image as PILImage
from traffic_light_msg.msg import TrafficLightMsg

class TrafficLightDataset(Dataset):
    def __init__(self, annotations_dir, data_dir, transform=None):
        self.annotations = []
        self.transform = transform
        self.data_dir = data_dir
        
        for clip_folder in os.listdir(annotations_dir):
            csv_path = os.path.join(annotations_dir, clip_folder, 'frameAnnotationsBOX.csv')
            if os.path.exists(csv_path):
                df = pd.read_csv(csv_path, sep=';')
                for _, row in df.iterrows():
                    img_name = row['Filename'].split('/')[-1]
                    clip_name = row['Origin file'].split('/')[-2]
                    img_path = os.path.join(data_dir, clip_name, 'frames', img_name)
                    if os.path.exists(img_path):
                        self.annotations.append({
                            'img_path': img_path,
                            'label': row['Annotation tag'],
                            'bbox': [
                                row['Upper left corner X'],
                                row['Upper left corner Y'],
                                row['Lower right corner X'],
                                row['Lower right corner Y']
                            ]
                        })
        
        self.class_to_idx = {cls: idx for idx, cls in enumerate(set([ann['label'] for ann in self.annotations]))}
    
    def __len__(self):
        return len(self.annotations)
    
    def __getitem__(self, idx):
        ann = self.annotations[idx]
        image = PILImage.open(ann['img_path']).convert('RGB')
        
        bbox = ann['bbox']
        label = ann['label']
        
        x1, y1, x2, y2 = bbox
        h, w = y2 - y1, x2 - x1
        center_x, center_y = (x1 + x2) / 2, (y1 + y2) / 2
        
        expand_ratio = 1.5
        new_h, new_w = int(h * expand_ratio), int(w * expand_ratio)
        new_x1 = max(0, int(center_x - new_w / 2))
        new_y1 = max(0, int(center_y - new_h / 2))
        new_x2 = min(image.width, int(center_x + new_w / 2))
        new_y2 = min(image.height, int(center_y + new_h / 2))
        
        cropped_img = image.crop((new_x1, new_y1, new_x2, new_y2))
        
        if self.transform:
            cropped_img = self.transform(cropped_img)
        
        return cropped_img, self.class_to_idx[label], torch.tensor(bbox), label


class CamNode(Node):
    def __init__(self):
        super().__init__('cam_node')
        self.publisher_ = self.create_publisher(Image, '/perception/traffic_light_recognition/traffic_light_image', 10)
        self.debug_publisher_ = self.create_publisher(TrafficLightMsg, '/perception/traffic_light_recognition/traffic_light_data', 10)
        self.timer = self.create_timer(1.0, self.timer_callback)
        self.cv_bridge = CvBridge()
        
        # 初始化数据集
        annotations_dir = '/home/casia/jupyter_notebooks/Annotations/Annotations/dayTrain'
        data_dir = '/home/casia/jupyter_notebooks/dayTrain/dayTrain'
        self.dataset = TrafficLightDataset(annotations_dir, data_dir, transform=transforms.ToTensor())
        
        self.get_logger().info(f"数据集大小: {len(self.dataset)}")
        self.get_logger().info(f"类别: {list(self.dataset.class_to_idx.keys())}")
        
        # 添加参数来控制是否使用调试模式
        self.declare_parameter('debug_mode', False)
        
    def timer_callback(self):
        # 随机选择一个样本
        idx = random.randint(0, len(self.dataset) - 1)
        image, class_idx, bbox, label = self.dataset[idx]
        
        # 将PyTorch张量转换为OpenCV图像
        image_np = (image.permute(1, 2, 0).numpy() * 255).astype('uint8')
        
        # 创建Image消息
        img_msg = self.cv_bridge.cv2_to_imgmsg(image_np, encoding="rgb8")
        
        # 发布Image消息
        self.publisher_.publish(img_msg)
        
        # 如果在调试模式下，发布TrafficLightMsg
        if self.get_parameter('debug_mode').value:
            debug_msg = TrafficLightMsg()
            debug_msg.image = img_msg
            debug_msg.label = label
            self.debug_publisher_.publish(debug_msg)
            self.get_logger().info(f'Published debug image with label: {label}')
        else:
            self.get_logger().info('Published image')

def main(args=None):
    rclpy.init(args=args)
    cam_node = CamNode()
    rclpy.spin(cam_node)
    cam_node.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()
